### ISSUE TITLE
Heart Rate Core dev update?

### ISSUE THREAD

**Issue by xuv** [matt-porter/BioSportHRService#1]
Hey, 

Searching for code examples on how to implement the Heart Rate Core SDK by Intel, I stumbled on your project files. Did you go any further with this? What stopped you? Would you like to exchange some info about this SDK?

Thx

**Comment by matt-porter** [comment]
Hi,

I managed to get as far as getting HR information from the headphones (simply logging it to the console; no Android integration), but they didn't work with my phone anyway, so I stopped development on this. 

I can have a look later and see if I can dig out what I did get working though, if it would be helpful.

**Comment by xuv** [comment]
Hello

Thanks for the quick response. And yes, if you can get a hand on your tests files again, that would be of great help. 

We're looking into using those headphones for an art project. And our dev knowledge is quite limited at this moment. So anything that could help us kickstart our project would be great. 

Any reason why the headphones did not work with your phone?

**Comment by matt-porter** [comment]
I don't know why they don't work with my phone, possibly a low power output on the headphone socket though. They do work with every other phone I tried them with though (HTC Desire 518, Nexus 5, Moto G (4g)).

**Comment by xuv** [comment]
Yes. That's what I'm thinking also. Tried with Nexus 5X and Pixel phone and we get audio signal. 


On March 22, 2017 8:38:34 AM EDT, Matt Porter <notifications@github.com> wrote:
>I don't know why they don't work with my phone, possibly a low power
>output on the headphone socket though. They do work with every other
>phone I tried them with though (HTC Desire 518, Nexus 5, Moto G (4g)).
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/matt-porter/BioSportHRService/issues/1#issuecomment-288385823

-- 
@xuv ~ http://xuv.be
Sent from mobile tracking device. 
Please excuse typos and brevity.

**Comment by matt-porter** [comment]
Yep - audio signal is what it looks like. I was half thinking about trying to reverse engineer that signal, before I discovered the Intel SDK.

I found the code, not sure what state it's in, but I'll add it to the repo now

**Comment by xuv** [comment]
Ho that's awesome. Perfect timing also. :)
Btw, if you look into reverse engineer the audio signal one day, here is something that might help you:
 
![biospost-audio-signal](https://cloud.githubusercontent.com/assets/192539/24313262/85efa78c-10b2-11e7-9646-f9757ee167de.jpeg)

**Comment by matt-porter** [comment]
Thanks :) I don't remember finding any information like that when I was looking at this; would certainly have helped!

**Comment by xuv** [comment]
And btw, thanks so much for your generosity. This code is really helpful. I'm looking into it at the moment. Something not working yet in the library linkage. Android Studio complained about some .iml files missing. Though they are in their folders. But thanks anyway. This is awesome.

**Comment by matt-porter** [comment]
No worries. Good luck for your project; let me know how it goes

**Comment by xuv** [comment]
Thx. I will. 

Since you asked ;) I'm looking into the permissions. I think they changed the way they handle those since you wrote your code. So now, we need to ask for RECORD_AUDIO permission. :)

**Comment by xuv** [comment]
@matt-porter Just to keep you updated.
I did not get your app to run. But it did help me understand how to set up and start the Heart Rate Core. I have now a simpler app running and getting the data. I basically did not need (or don't think I need) the fitness.sensor android stuff. Although your idea to couple both of them is quite genius.

On the other hand, I can run the code on an armv7 chip phone, no problem. But on an armv8 phone, it produces an error from the SDK in the form of:
```E/linker: lib/arm/libalgo_wrap.so: has text relocations```
By any chance would you understand what that means? 

I must add that the SDK has 3 `/jni` subfolders named `armeabi`, `armeabi-v7a` and `x86`. Do I miss one for the armv8 architecture or should it pick up from one of those?

Thx.

**Comment by matt-porter** [comment]
Nice one. Sorry, no idea about your error; maybe the Intel SDK wasn't written for arm8?

On 31 Mar 2017 5:08 pm, Julien Deswaef <notifications@github.com> wrote:

@matt-porter<https://github.com/matt-porter> Just to keep you updated.
I did not get your app to run. But it did help me understand how to set up and start the Heart Rate Core. I have now a simpler app running and getting the data. I basically did not need (or don't think I need) the fitness.sensor android stuff. Although your idea to couple both of them is quite genius.

On the other hand, I can run the code on an armv7 chip phone, no problem. But on an armv8 phone, it produces an error from the SDK in the form of:
E/linker: lib/arm/libalgo_wrap.so: has text relocations
By any chance would you understand what that means?

Thx.

**Comment by xuv** [comment]
Yes. That's my guess also. Though I though armv8 and armv7 were compatible...

**Comment by dfl** [comment]
I have been working on reverse-engineering the sensor data from the encoded audio signal... banging my head against the wall though :P Last year I compiled an iPhone app but the SDK seemed to be totally unsupported.

**Comment by xuv** [comment]
@dfl Indeed, SDK is totally unsuported. And the feedback I have is that it's never going to be supported again. Nor will they release source code or anything for us to try to get it to work on our own.

**Comment by dfl** [comment]
someone at [dsp.stackexchange](https://dsp.stackexchange.com/questions/47685/mystery-signal-fsk-bpsk) helped me out a bit... apparently it's not Manchester encoded nor is it 4-8kHz square wave :P 
![](https://www.evernote.com/shard/s262/sh/96e90924-3e65-4611-817a-50b4d7b80b4a/4ffbd21902edc540/res/eee52e0a-d8f3-4ee4-a25a-bcc6b000f6e4/skitch.png)
Looks like I have the binary data... now to convert it into a PCM waveform of the PPG signal. I think I'm close to a solution!

**Comment by dfl** [comment]
So now that I've decoded the differential Manchester coding... I am getting the supposed PCM bytes but I have no clue what they mean... whether big/little endian, mu-law, etc. I'm expecting the output should be the PPG pulse waveform. Anyone care to take a crack at the puzzle? https://www.dropbox.com/s/om7o7t3hty5l7nk/decoded.txt?dl=0

**Comment by weakiwi** [comment]
@dfl Do you have any imporvement on reverse engineering?
Like a code or demo

**Comment by dfl** [comment]
No, sadly I gave up. I suspect it might only give packets of heart rate data, not the actual PPG wave in realtime. That and It's basically it's impossible to crack without the specs :(

**Comment by weakiwi** [comment]
@dfl  Thanks for your contribution on this project!
Would you might to share example code with me? I wonder check if it can work with my device(the runkeeper app give up supporting devices and most of the android phone doesn't support it either)

**Comment by dfl** [comment]
@weakiwi I emailed you my matlab code to your qq.com address

**Comment by weakiwi** [comment]
@dfl I find something interesting by decompiling intel HR sdk. It seems the protocol is performTek.
And there is a .[performTek header file](https://github.com/DemonTu/ALL_HeartRate_CC1310/blob/461c6409949d7e9f7ecdc7f96f95b2ec46c0fd09/IAR/worksplace/HeartRate/Application/userApp/UserHeartBeatApp.h)
